### PR TITLE
Proper TReflectionGameTab.Open function added & Widget left-click uptext check

### PR DESCRIPTION
**Pull request by NicklasSoderberg**
Proper TReflectionGameTab.Open function added & Widget left-click uptext check
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055676%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055741%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055800%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055821%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055839%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055862%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055897%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055907%22%2C%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055919%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/internal/Login.simba%20173%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055800%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22GetColor%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A22%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/internal/Login.simba%3AL234-248%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/internal/Login.simba%2073%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055676%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Mouse_Left%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A21%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/internal/Login.simba%3AL25-94%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/widgets/gametabs/Gametabs.simba%2025%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055919%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%5E%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A23%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/widgets/gametabs/Gametabs.simba%3AL30-61%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/internal/Login.simba%20199%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055839%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22not%2C%20Login_%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A22%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/internal/Login.simba%3AL252-268%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/internal/Login.simba%20191%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055821%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22GetColor%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A22%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/internal/Login.simba%3AL252-268%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/widgets/gametabs/Gametabs.simba%2016%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055907%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Boolean%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A23%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/widgets/gametabs/Gametabs.simba%3AL9-23%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/internal/Login.simba%20111%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055741%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Integer%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A21%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/internal/Login.simba%3AL25-94%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/widgets/Bank.simba%205%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055862%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Boolean%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A23%3A00Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/widgets/Bank.simba%3AL280-308%22%7D%2C%20%22Pull%20beaee7ce165b0d18778d00a6ef5f225add703a11%20lib/widgets/gametabs/Gametabs.simba%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Elfyyy/OSR-Reflection/pull/54%23discussion_r26055897%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22No%20need%20to%20reorder%20these%22%2C%20%22created_at%22%3A%20%222015-03-09T17%3A23%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/4069146%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Elfyyy%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/widgets/gametabs/Gametabs.simba%3AL9-23%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/internal/Login.simba 73'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055676'>File: lib/internal/Login.simba:L25-94</a></b>
- <a href='https://github.com/Elfyyy'></a> Mouse_Left
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/internal/Login.simba 111'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055741'>File: lib/internal/Login.simba:L25-94</a></b>
- <a href='https://github.com/Elfyyy'></a> Integer
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/internal/Login.simba 173'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055800'>File: lib/internal/Login.simba:L234-248</a></b>
- <a href='https://github.com/Elfyyy'></a> GetColor
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/internal/Login.simba 191'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055821'>File: lib/internal/Login.simba:L252-268</a></b>
- <a href='https://github.com/Elfyyy'></a> GetColor
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/internal/Login.simba 199'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055839'>File: lib/internal/Login.simba:L252-268</a></b>
- <a href='https://github.com/Elfyyy'></a> not, Login_
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/widgets/Bank.simba 5'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055862'>File: lib/widgets/Bank.simba:L280-308</a></b>
- <a href='https://github.com/Elfyyy'></a> Boolean
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/widgets/gametabs/Gametabs.simba 7'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055897'>File: lib/widgets/gametabs/Gametabs.simba:L9-23</a></b>
- <a href='https://github.com/Elfyyy'></a> No need to reorder these
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/widgets/gametabs/Gametabs.simba 16'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055907'>File: lib/widgets/gametabs/Gametabs.simba:L9-23</a></b>
- <a href='https://github.com/Elfyyy'></a> Boolean
- [ ] <a href='#crh-comment-Pull beaee7ce165b0d18778d00a6ef5f225add703a11 lib/widgets/gametabs/Gametabs.simba 25'></a> &nbsp;<b><a href='https://github.com/Elfyyy/OSR-Reflection/pull/54#discussion_r26055919'>File: lib/widgets/gametabs/Gametabs.simba:L30-61</a></b>
- <a href='https://github.com/Elfyyy'></a> ^

<a href='https://www.codereviewhub.com/Elfyyy/OSR-Reflection/pull/54?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/Elfyyy/OSR-Reflection/pull/54?approve=1'></a>&nbsp;<a href='https://github.com/Elfyyy/OSR-Reflection/pull/54'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by Elfyyy** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/Elfyyy/OSR-Reflection/pull/54'>Refresh</a>.<a href='#crh-update-none'></a>

File: lib/internal/Login.simba [hunk 1]
```diff
@@ -48,50 +25,69 @@ begin
...
-    if (CountColor(Login_Text_Yellow, Login_PasswordBox.X1,
-      Login_PasswordBox.Y1, Login_PasswordBox.X2,
-      Login_PasswordBox.Y2) > 10) then
-        Exit(Login_PasswordBox_Byte);
-    Wait(Random(50, 100));
-  until (GetTickCount64 > TimeOut);
+    If not Reflect.Login.AtLogin then
+      Exit(False);
+    Inc(Tries);
+    Reflect.Mouse.Move(ClickSwitchButton, mouse_Left);
```
Comment (KyleHunter) [inline lib/internal/Login.simba]: Mouse_Left

File: lib/internal/Login.simba [hunk 2]
```diff
@@ -48,50 +25,69 @@ begin
...
+  Reflect.Mouse.Move(Reflect.Math.GridBox(I + 1, 4, 17, 87, 18, 93, 24, Point(242, 70)),
+    Mouse_Left);
+  T.Start;
+  while (Reflect.Login.AtWorldScreen) and (T.ElapsedTime < 5000) do
+    Wait(RandomRange(100, 200));
+  Result := (not Reflect.Login.AtWorldScreen);
 end;
 
-function TReflectionLogin.LoginState: integer;
+function TReflectionLogin.GetLoginState: integer;
```
Comment (KyleHunter) [inline lib/internal/Login.simba]: Integer

File: lib/internal/Login.simba [hunk 3]
```diff
@@ -205,14 +234,14 @@ var
   CurState, Tries: Integer;
   T, TLogged: TReflectTimer;
 begin
-  If Reflect.Login.GetState = 30 then
+  If (Reflect.Login.GetClientState = 30) and (getColor(427, 489) = 16777215) then
```
Comment (KyleHunter) [inline lib/internal/Login.simba]: GetColor

File: lib/internal/Login.simba [hunk 4]
```diff
@@ -223,15 +252,16 @@ begin
         Reflect.Mouse.Move(330, 315, 483, 360, Mouse_Left);
         TLogged.Start;
         repeat
-          Result := (Reflect.Login.LoginState = -1);
+          Result := (Reflect.Login.GetClientState = 30) and (getColor(427, 489) = 16777215);
```
Comment (KyleHunter) [inline lib/internal/Login.simba]: GetColor

File: lib/internal/Login.simba [hunk 5]
```diff
@@ -223,15 +252,16 @@ begin
...
           Wait(RandomRange(100, 200));
-        until(TLogged.ElapsedTime > 8000)or(Result);
+        until(TLogged.ElapsedTime > 3000)or(Result);
       end;
       else
         Wait(RandomRange(100, 200));
     end;
     If Not (InIntArray([LOGIN_Connecting, LOGIN_ClickToPlay], CurState))then
```
Comment (KyleHunter) [inline lib/internal/Login.simba]: not, Login_

File: lib/widgets/Bank.simba [hunk 6]
```diff
@@ -280,15 +280,28 @@ begin
   BankWidget.Free;
 end;
 
-procedure TReflectionBank.Close;
+function TReflectionBank.Close: boolean;
```
Comment (KyleHunter) [inline lib/widgets/Bank.simba]: Boolean

File: lib/widgets/gametabs/Gametabs.simba [hunk 7]
```diff
@@ -9,14 +9,14 @@ begin
     Widget_IgnoreList_Container, Widget_Logout_Container,
     Widget_Options_Container, Widget_Emotes_Container,
     Widget_MusicPlayer_Container, Widget_CombatOptions_Container,
-    Widget_Stats_Container, Widget_QuestList_Container,
-    Widget_Inventory_Container, Widget_WornEquipment_Container,
-    Widget_Prayer_Container, Widget_Magic_Container
+    Widget_Stats_Container, Widget_Inventory_Container,
```
Comment (KyleHunter) [inline lib/widgets/gametabs/Gametabs.simba:12]: No need to reorder these

File: lib/widgets/gametabs/Gametabs.simba [hunk 8]
```diff
@@ -9,14 +9,14 @@ begin
...
+    Widget_WornEquipment_Container, Widget_Prayer_Container,
+    Widget_Magic_Container, Widget_QuestList_Container
   ];
   Gametabs := [Gametab_ClanChat, Gametab_FriendsList, Gametab_IgnoreList,
     Gametab_Logout, Gametab_Options, Gametab_Emotes, Gametab_MusicPlayer,
-    Gametab_CombatOptions, Gametab_Stats, Gametab_QuestList, Gametab_Inventory,
-    Gametab_WornEquipment, Gametab_Prayer, Gametab_Magic
+    Gametab_CombatOptions, Gametab_Stats, Gametab_Inventory,
+    Gametab_WornEquipment, Gametab_Prayer, Gametab_Magic, Gametab_QuestList
```
Comment (KyleHunter) [inline lib/widgets/gametabs/Gametabs.simba:19]: Boolean

File: lib/widgets/gametabs/Gametabs.simba [hunk 9]
```diff
@@ -30,14 +30,31 @@ begin
   end;
 end;
 
-procedure TReflectionGametab.Open(TabNum: Integer);
+function TReflectionGametab.Open(TabNum: Integer): boolean;
```
Comment (KyleHunter) [inline lib/widgets/gametabs/Gametabs.simba]: ^

**Pull request merged**